packages/webview_flutter/webview_flutter_wkwebview/lib/src/v4/src/webkit_navigation_delegate.dart

```diff
@@ -0,0 +1,158 @@
+class WebKitNavigationDelegate extends PlatformNavigationDelegate {
+  /// Constructs a [WebKitNavigationDelegate].
+  WebKitNavigationDelegate(
+    super.params, {
+    @visibleForTesting WebKitProxy webKitProxy = const WebKitProxy(),
+  }) : super.implementation() {
+    final WeakReference<WebKitNavigationDelegate> weakThis =
+        WeakReference<WebKitNavigationDelegate>(this);
+    navigationDelegate = webKitProxy.createNavigationDelegate(
+      didFinishNavigation: (WKWebView webView, String? url) {
+        if (weakThis.target?._onPageFinished != null) {
+          weakThis.target!._onPageFinished!(url ?? '');
+        }
+      },
+      didStartProvisionalNavigation: (WKWebView webView, String? url) {
+        if (weakThis.target?._onPageStarted != null) {
+          weakThis.target!._onPageStarted!(url ?? '');
+        }
+      },
+      decidePolicyForNavigationAction: (
+        WKWebView webView,
+        WKNavigationAction action,
+      ) async {
+        if (weakThis.target?._onNavigationRequest != null) {
+          final bool allow = await weakThis.target!._onNavigationRequest!(
+            url: action.request.url,
+            isForMainFrame: action.targetFrame.isMainFrame,
+          );
+          return allow
+              ? WKNavigationActionPolicy.allow
+              : WKNavigationActionPolicy.cancel;
+        }
+        return WKNavigationActionPolicy.allow;
+      },
+      didFailNavigation: (WKWebView webView, NSError error) {
+        if (weakThis.target?._onWebResourceError != null) {
+          weakThis.target!._onWebResourceError!(
+            WebKitWebResourceError._(error),
+          );
+        }
+      },
+      didFailProvisionalNavigation: (WKWebView webView, NSError error) {
+        if (weakThis.target?._onWebResourceError != null) {
+          weakThis.target!._onWebResourceError!(
+            WebKitWebResourceError._(error),
+          );
+        }
+      },
+      webViewWebContentProcessDidTerminate: (WKWebView webView) {
+        if (weakThis.target?._onWebResourceError != null) {
+          weakThis.target!._onWebResourceError!(
+            WebKitWebResourceError._(
+              const NSError(
+                code: WKErrorCode.webContentProcessTerminated,
+                // Value from https://developer.apple.com/documentation/webkit/wkerrordomain?language=objc.
+                domain: 'WKErrorDomain',
+                localizedDescription: '',
+              ),
+            ),
+          );
+        }
+      },
+    );
+  }
+
+  /// WebKit class that handles navigation changes and tracking navigation
+  /// requests.
+  late final WKNavigationDelegate navigationDelegate;
+
+  void Function(String url)? _onPageFinished;
+  void Function(String url)? _onPageStarted;
+  void Function(int progress)? _onProgress;
+  void Function(WebResourceError error)? _onWebResourceError;
+  FutureOr<bool> Function({required String url, required bool isForMainFrame})?
+      _onNavigationRequest;
```

Comment:
In defense of this pattern:

I wanted setters for the callbacks methods because I wanted a consistent way for a platform to indicate it doesn't support a feature.

For example, if Android doesn't support `WebViewController.callSomeMethod()`, it throws an `UnimplementedError` by default. By using setters for callback methods, the default response for a callback method is to throw an `UnimplementedError`. I didn't know a way to do this with a class that takes the callback methods as constructor parameters. 

Also note that the app facing package will probably use fields:

```dart
class NavigationDelegate {
  final Function(String) onPageFinished;
}
```

But the platform classes will use setters so they can throw when a callback is not supported.

Reply:
> By using setters for callback methods, the default response for a callback method is to throw an `UnimplementedError`.

I think I'm missing something about the implementation; where does that happen? It looks in the code above this like the default would be a silent no-op for a null method.

Reply:
It's done in the platform interface class: https://github.com/flutter/plugins/blob/main/packages/webview_flutter/webview_flutter_platform_interface/lib/v4/src/platform_navigation_delegate.dart#L63

`PlatformNavigationDelegate` has to be extended, so when a new callback method is added (e.g. `setOnPageRedirected`) platform implementations would throw an `UnimplementedError` by default.